tardis/analysis_tools/line_ID.py

```diff
@@ -0,0 +1,274 @@
+
+    def _reset_cache(self):
+        self._reset_lam_min()
+        self._reset_lam_max()
+        self._reset_derived_quantities()
+
+    def _reset_lam_min(self):
+        self._lam_min = None
+
+    def _reset_lam_max(self):
+        self._lam_max = None
+
+    def _reset_derived_quantities(self):
+        self._line_mask = None
+        self._lam_in = None
+        self._lines_info_unique = None
+        self._lines_count = None
+        self._lines_ids = None
+        self._lines_ids_unique = None
+
+    @property
+    def lam_min(self):
+        if self._lam_min is None:
+            raise ValueError("lam_min not set")
+        return self._lam_min
+
+    @lam_min.setter
+    def lam_min(self, val):
+        self._reset_derived_quantities()
+        try:
+            self._lam_min = val.to(u.AA)
+        except AttributeError:
+            self._lam_min = val * u.AA
+
+    @property
+    def lam_max(self):
+        if self._lam_max is None:
+            raise ValueError("lam_max is not set")
+        return self._lam_max
+
+    @lam_max.setter
+    def lam_max(self, val):
+        self._reset_derived_quantities()
+        try:
+            self._lam_max = val.to(u.AA)
+        except AttributeError:
+            self._lam_max = val * u.AA
+
+    @property
+    def lam_in(self):
+        if self._lam_in is None:
+            self._lam_in = u.Quantity(
+                self.data.packets_df.last_line_interaction_in_nu, u.Hz
+            ).to(u.AA, equivalencies=u.spectral())
+        return self._lam_in
+
+    @property
+    def line_mask(self):
+        if self._line_mask is None:
+            self._line_mask = (self.lam_in >= self.lam_min) * (
+                self.lam_in <= self.lam_max
+            )
+        return self._line_mask
+
+    @property
+    def lines_ids(self):
+        if self._lines_ids is None:
+            ids = self.data.packets_df.last_line_interaction_in_id[
+                self.line_mask
+            ].values
+            self._lines_ids = self.data.lines_df.iloc[ids].index
+        return self._lines_ids
+
+    @property
+    def lines_ids_unique(self):
+        if self._lines_ids_unique is None:
+            self._lines_ids_unique = np.unique(self.lines_ids)
+        return self._lines_ids_unique
+
+    @property
+    def lines_info_unique(self):
+        if self._lines_info_unique is None:
+            self._lines_info_unique = self.data.lines_df.loc[
+                self.lines_ids_unique
+            ]
+        return self._lines_info_unique
+
+    @property
+    def lines_count(self):
+        if self._lines_count is None:
+            counts = np.bincount(self.lines_ids)
+            self._lines_count = counts[counts > 0]
+        return self._lines_count
+
+    def identify(self, lam_min, lam_max):
+        self.lam_min = lam_min
+        self.lam_max = lam_max
+
+    def plot_summary(
+        self, nlines=None, lam_min=None, lam_max=None, output_filename=None
```

Comment:
IMO we should separate out line contributions file writing from `plot_summary()` by creating another method like `save_summary()`. It's not clear to me why it has to be done in plotting function - let me know if I'm missing something